File: claude-code-agents-typescript/src/models/crawl-target.ts
// src/models/crawl-target.ts — Crawl target and plan models for campaign orchestration
import { z } from 'zod';
import type {
  CampaignId,
  QualityValue,
  SpiderName,
  USD,
  Url,
} from '../types.js';
import {
  assertNever,
  toCampaignId,
  toQualityValue,
  toSpiderName,
  toUrl,
} from '../types.js';

// ── Page Type (Discriminated Union) ─────────────────────────────
export const PAGE_TYPES = [
  'doc',
  'research',
  'news',
  'api',
  'legal',
  'product',
  'plugin_spec',
  'sdk_ref',
] as const;

export type PageType = (typeof PAGE_TYPES)[number];

export function isPageType(value: string): value is PageType {
  return (PAGE_TYPES as readonly string[]).includes(value);
}

export function toPageType(value: string): PageType {
  const lower = value.trim().toLowerCase();
  if (isPageType(lower)) return lower;
  return 'doc';
}

export function describePageType(pt: PageType): string {
  switch (pt) {
    case 'doc':
      return 'Documentation page';
    case 'research':
      return 'Research paper or article';
    case 'news':
      return 'News or announcement';
    case 'api':
      return 'API reference';
    case 'legal':
      return 'Legal or policy page';
    case 'product':
      return 'Product page';
    case 'plugin_spec':
      return 'Plugin specification';
    case 'sdk_ref':
      return 'SDK reference';
    default:
      return assertNever(pt);
  }
}

// ── Zod Schemas ─────────────────────────────────────────────────
export const CrawlTargetSchema = z.object({
  url: z.string().url(),
  spiderName: z.string().default('generic'),
  maxPages: z.number().int().min(1).max(10000).default(50),
  priority: z.number().int().min(0).max(10).default(0),
  allowedDomains: z.array(z.string()).default([]),
  pageTypeHint: z.enum(PAGE_TYPES).optional(),
});

export type CrawlTargetInput = z.input<typeof CrawlTargetSchema>;

export interface CrawlTarget {
  readonly url: Url;
  readonly spiderName: SpiderName;
  readonly maxPages: number;
  readonly priority: number;
  readonly allowedDomains: readonly string[];
  readonly pageTypeHint: PageType | undefined;
}

export function createCrawlTarget(input: CrawlTargetInput): CrawlTarget {
  const parsed = CrawlTargetSchema.parse(input);
  return {
    url: toUrl(parsed.url),
    spiderName: toSpiderName(parsed.spiderName),
    maxPages: parsed.maxPages,
    priority: parsed.priority,
    allowedDomains: parsed.allowedDomains,
    pageTypeHint: parsed.pageTypeHint,
  };
}

export function effectiveDomains(target: CrawlTarget): readonly string[] {
  if (target.allowedDomains.length > 0) return target.allowedDomains;
  try {
    const parsed = new URL(target.url);
    return [parsed.hostname];
  } catch {
    return [];
  }
}

// ── Crawl Plan ──────────────────────────────────────────────────
export const CrawlPlanSchema = z.object({
  targets: z.array(CrawlTargetSchema).default([]),
  totalBudgetUsd: z.number().positive().max(1000).default(5.0),
  maxIterations: z.number().int().min(1).max(20).default(3),
  concurrency: z.number().int().min(1).max(10).default(2),
  qualityThreshold: z.number().min(0).max(1).default(0.8),
});

export type CrawlPlanInput = z.input<typeof CrawlPlanSchema>;

export interface CrawlPlan {
  readonly id: CampaignId;
  readonly targets: readonly CrawlTarget[];
  readonly totalBudgetUsd: USD;
  readonly maxIterations: number;
  readonly concurrency: number;
  readonly qualityThreshold: QualityValue;
}

export function createCrawlPlan(input: CrawlPlanInput): CrawlPlan {
  const parsed = CrawlPlanSchema.parse(input);
  return {
    id: toCampaignId(crypto.randomUUID()),
    targets: parsed.targets.map(createCrawlTarget),
    totalBudgetUsd: parsed.totalBudgetUsd as USD,
    maxIterations: parsed.maxIterations,
    concurrency: parsed.concurrency,
    qualityThreshold: toQualityValue(parsed.qualityThreshold),
  };
}

export function totalMaxPages(plan: CrawlPlan): number {
  return plan.targets.reduce((sum, t) => sum + t.maxPages, 0);
}

export function sortedTargets(plan: CrawlPlan): readonly CrawlTarget[] {
  return [...plan.targets].sort((a, b) => b.priority - a.priority);
}

File: claude-code-agents-python/src/models/crawl_target.py
"""Crawl target and plan models for campaign orchestration."""

from __future__ import annotations

from enum import Enum
from typing import Optional

from pydantic import BaseModel, Field, HttpUrl

__all__ = ["CrawlTarget", "CrawlPlan", "PageType"]


class PageType(str, Enum):
    """Classification of crawled page types."""

    DOC = "doc"
    RESEARCH = "research"
    NEWS = "news"
    API = "api"
    LEGAL = "legal"
    PRODUCT = "product"
    PLUGIN_SPEC = "plugin_spec"
    SDK_REF = "sdk_ref"


class CrawlTarget(BaseModel):
    """A single URL target for crawling."""

    url: str = Field(..., description="The URL to crawl")
    spider_name: str = Field(
        default="generic", description="Name of the Scrapy spider to use"
    )
    max_pages: int = Field(default=50, ge=1, le=10000, description="Maximum pages to crawl")
    priority: int = Field(
        default=0, ge=0, le=10, description="Priority level (0=lowest, 10=highest)"
    )
    allowed_domains: list[str] = Field(default_factory=list, description="Restrict crawl to these domains")
    page_type_hint: Optional[PageType] = Field(
        default=None, description="Hint for expected page type"
    )

    def effective_domains(self) -> list[str]:
        """Return allowed domains, falling back to domain extracted from URL."""
        if self.allowed_domains:
            return self.allowed_domains
        from urllib.parse import urlparse

        parsed = urlparse(self.url)
        if parsed.hostname:
            return [parsed.hostname]
        return []


class CrawlPlan(BaseModel):
    """A plan describing a full crawl campaign."""

    targets: list[CrawlTarget] = Field(
        default_factory=list, description="List of crawl targets"
    )
    total_budget_usd: float = Field(
        default=5.0, gt=0, le=1000, description="Total budget in USD for API calls"
    )
    max_iterations: int = Field(
        default=3, ge=1, le=20, description="Maximum improvement iterations"
    )
    concurrency: int = Field(default=2, ge=1, le=10, description="Parallel crawl tasks")
    quality_threshold: float = Field(
        default=0.8, ge=0.0, le=1.0, description="Minimum quality score to stop iterating"
    )

    @property
    def total_max_pages(self) -> int:
        """Sum of max_pages across all targets."""
        return sum(t.max_pages for t in self.targets)

    def sorted_targets(self) -> list[CrawlTarget]:
        """Return targets sorted by priority descending."""
        return sorted(self.targets, key=lambda t: t.priority, reverse=True)
